nomos-services/membership/src/backends/sdp.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::HashMap;
+
+use nomos_sdp::backends::FinalizedBlockEvent;
+use nomos_sdp_core::{DeclarationUpdate, ProviderInfo, ServiceType};
+
+use super::{MembershipBackend, MembershipBackendSettings, MembershipEntry, SnapshotSettings};
+
+pub struct SdpMembershipBackend {
+    settings: MembershipBackendSettings,
+    // todo: storage trait instead of deque in memory
+    membership: Vec<MembershipEntry>,
+    current_data: HashMap<ServiceType, HashMap<ProviderInfo, DeclarationUpdate>>,
+}
+
+#[async_trait::async_trait]
+impl MembershipBackend for SdpMembershipBackend {
+    fn init(settings: MembershipBackendSettings) -> Self {
+        Self {
+            settings,
+            membership: Vec::new(),
+            current_data: HashMap::new(),
+        }
+    }
+
+    async fn get_snapshot_at(
+        &self,
+        service_type: ServiceType,
+        index: i32,
+    ) -> Result<HashMap<ProviderInfo, DeclarationUpdate>, overwatch::DynError> {
```

Comment:
Would it make sense here to return only `ProviderId + Declaration` instead of `ProviderInfo + DeclarationUpdate`, or even better - `ProviderId + Locators`. In this case provider id is `Ed25519_Public_Key` that we'll use to derive `PeerId` from where required, and locators are the addresses to try and connect to. I think other information is not needed by the Membership service users.

Reply:
Great question, I was not sure about the concrete type (and error type) yet, as I did not fully implemented membership service. 

What does membership service needs relies on what do we need to create a snapshot.  I was thinking that maybe we will need to also fetch service params and remove the expired declarations in order for active providers to be correct in each snapshot.

The idea was to maybe use ServicesAdapter from Sdp to fetch service parameters and use the info from ProviderInfo to expiry old ones. We could also rely on the sdp to provide only up-to-date information, depending on how we implement garbage collection. 

What is your opinion on this? I think if we don't have to check the expiration, then I 200% agree that only ProviderId and Locators is enough.

Reply:
For DA and Blend normal operation we'll need to stream the latest active provider ids and their locators. For DA, we'll need historic data based on BlockNumber. Executor network most likely will need only the latest data.
`get_snapshot_at` could stay as is by providing only ids and locators, and then we still would require a message to subscribe to latest provider ids that became active, or got removed.

Reply:
Great, I would add subscription to the API as well then. Thank you